src/osdc/Objecter.cc

```diff
@@ -2316,6 +2319,52 @@ void Objecter::_op_submit(Op *op, shunique_lock<ceph::shared_mutex>& sul, ceph_t
   bool check_for_latest_map = _calc_target(&op->target, nullptr)
     == RECALC_OP_TARGET_POOL_DNE;
 
+  if (op->tid == 0)
+    op->tid = ++last_tid;
+  if (objecter_fastfail_timeout  > timespan(0)) {
+    unique_lock fl(ffpgs.lock);
+    if (ffpgs.pgs.find(op->target.pgid) != ffpgs.pgs.end()) {
+      fl.unlock();
+      ldout(cct, 10) << "fastfail pg: " << op->target.pgid << " to list" << dendl;
+      // before cancelling check if the pg is still inactive
+      // TODO: find a better way to determine if a pg is inactive
+      if(op->target.acting.size() < op->target.min_size) {
+        boost::asio::dispatch(service,
+				      [this, tid=op->tid]() {
+                ldout(cct, 10) << "Cancel op" << dendl;
+                op_cancel(tid, -ETIMEDOUT);
+              });
+      } else {
+        unique_lock fl(ffpgs.lock);
```

Comment:
if we've just cancelled fastfail, shouldn't we treat this like an op submitted without fastfail for the pg and handle it timing out in the same way as any new op?